Cache env so we don't have to spawn node on subsequent invokations

andreypopp opened the pull request:
This shaves around 200-300ms from `esy` and `esy CMD` subsequent invokations:
from ~500ms down to ~200ms.

We use the same mechanism to check mtimes as for build-eject caching: make but
probably can do better if we just check mtime manually in bash. This is left
for future improvements though.

*Update:* dropped `make` and rewrote mtime comparison in bash — now it's 100ms less.

```
% time ../esy/.bin/esy true
../esy/.bin/esy true  0.04s user 0.05s system 84% cpu 0.111 total
```

andreypopp committed:
Cache env so we don't have to spawn node on subsequent invokations

This shaves around 200-300ms from `esy` and `esy CMD` subsequent invokations:
from ~500ms down to ~200ms.

We use the same mechanism to check mtimes as for build-eject caching: make but
probably can do better if we just check mtime manually in bash. This is left
for future improvements though.

andreypopp committed:
Do not use make to check for ejected env/build

This shaves around ~100ms from cached env/build invokations.

jordwalke commented on .bin/esy:
```diff
@@ -27,30 +27,69 @@ SCRIPTDIR="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
 # stderr to nonblocking mode
 
 EJECT_PATH="$PWD/node_modules/.cache/esy"
+ENV_PATH="$PWD/node_modules/.cache/esy-env"
```
This is going to be great for editor/IDE integration. In fact, this can become a language/build-system agnostic convention.

cc @freebroccolo.

jordwalke commented:
🎉 Merge when you're ready!

andreypopp committed:
install: fix pkg store location when set via env

andreypopp committed:
esy: set env vars earlier

That's more consistent than relying on defaults in Node modules and in shell
wrapper.
